internal/core/metadata/rest_provisionwatcher.go

```diff
@@ -386,75 +387,82 @@ func restUpdateProvisionWatcher(
 	errorHandler errorconcept.ErrorHandler) {
 
 	defer r.Body.Close()
-	var from models.ProvisionWatcher
-	if err := json.NewDecoder(r.Body).Decode(&from); err != nil {
+	var to models.ProvisionWatcher
+	if err := json.NewDecoder(r.Body).Decode(&to); err != nil {
 		errorHandler.Handle(w, err, errorconcept.Common.InvalidRequest_StatusServiceUnavailable)
 		return
 	}
 
-	// Check if the provision watcher exists
-	// Try by ID
-	to, err := dbClient.GetProvisionWatcherById(from.Id)
-	if err != nil {
-		// Try by name
-		if to, err = dbClient.GetProvisionWatcherByName(from.Name); err != nil {
-			errorHandler.HandleOneVariant(
-				w,
-				err,
-				errorconcept.ProvisionWatcher.NotFoundByName,
-				errorconcept.Common.RetrieveError_StatusServiceUnavailable)
+	// explicitly do not check for errors here, we may have to lookup by name
+	from, _ := dbClient.GetProvisionWatcherById(to.Id)
+
+	byName, err := dbClient.GetProvisionWatcherByName(to.Name)
+	if err == nil {
+		// the only name collision we allow is same name and same ID,
+		// because that just means neither the ID or the name is going to change
+		// we have the guard for empty update IDs to allow for updating by name
+		if (byName.Name == to.Name && byName.Id != to.Id && to.Id != "") ||
+			(byName.Name == from.Name && byName.Id != from.Id && to.Id != "") {
+			errorHandler.Handle(w, err, errorconcept.NewProvisionWatcherDuplicateErrorConcept(byName.Id, to.Id))
```

Comment:
The changes in this file look good, but were a little hard for me to follow at first. Would we be able to move some of the logic into helper functions for better readability? Also, if we do it would help a little more if we can eliminate some of the side-effects. For example, the `if` statement could be maybe look something like:
```go
if (!identifiersAreValid(byName,to) || !identifiersAreValid(byName,from){
  	errorHandler.Handle(w, err, errorconcept.NewProvisionWatcherDuplicateErrorConcept(byName.Id, to.Id))
}
```
where the helper function `identifiersAreValid` help the readability at this level and the reader can dive into the details by going to that function.

As for the side-effects comment, here we are double using the `from` variable. In one instance we are storing the `ProvisionWatcher` retrieved from the database by ID. In the other we are using it as the base to which we apply updates, but if it is empty then we populate `from` with the data from `byName`. Could we possibly create a function which performs this logic and simply returns the "base" to which we can apply updates.

Reply:
Great suggestions. Addressed.